base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -190,45 +197,62 @@ fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, Mer
     Ok(difficulty)
 }
 
-/// Appends merge mining hash to a Monero block and returns the encoded Monero blocktemplate_blob
-pub fn append_merge_mining_tag(block: &mut MoneroBlock, hash: Hash) -> Result<String, MergeMineError> {
+/// Appends merge mining hash to a Monero block
+pub fn append_merge_mining_tag<T: AsRef<[u8]>>(block: &mut MoneroBlock, hash: T) -> Result<(), MergeMineError> {
+    if hash.as_ref().len() != Hash::len_bytes() {
+        return Err(MergeMineError::HashingError(format!(
+            "Expected source to be {} bytes, but it was {} bytes",
+            Hash::len_bytes(),
+            hash.as_ref().len()
+        )));
+    }
+    let hash = Hash::from_slice(hash.as_ref());
     let mm_tag = SubField::MergeMining(VarInt(0), hash);
     block.miner_tx.prefix.extra.0.push(mm_tag);
-    let serialized = serialize::<MoneroBlock>(&block);
-    Ok(hex::encode(&serialized).into())
+    Ok(())
 }
 
-/// Calculates the encoded Monero blockhashing_blob
-pub fn create_input_blob(
-    header: &MoneroBlockHeader,
-    tx_count: &u16,
-    tx_hashes: &Vec<[u8; 32]>,
-) -> Result<String, MergeMineError>
-{
-    let header = serialize::<MoneroBlockHeader>(header);
-    // Note count assumes the miner tx is included already
-    let mut count = serialize::<VarInt>(&VarInt(tx_count.clone() as u64));
-    let mut hashes = Vec::new();
-    for item in tx_hashes {
-        hashes.push(Hash::from(item.clone()));
-    }
-    let mut root = tree_hash(hashes)?;
+/// Creates a hex encoded Monero blockhashing_blob
+pub fn create_input_blob(block: &MoneroBlock) -> Result<String, MergeMineError> {
+    let tx_hashes = create_ordered_transaction_hashes_from_block(block);
+    create_input_blob_from_parts(&block.header, &tx_hashes)
+}
+
+pub fn create_ordered_transaction_hashes_from_block(block: &MoneroBlock) -> Vec<Hash> {
+    iter::once(block.miner_tx.hash())
+        .chain(block.tx_hashes.clone())
+        .collect()
+}
+
+/// Creates a hex encoded Monero blockhashing_blob
+fn create_input_blob_from_parts(header: &MoneroBlockHeader, tx_hashes: &[Hash]) -> Result<String, MergeMineError> {
+    let header = serialize::<MoneroBlockHeader>(&header);
     let mut encode = header;
-    encode.append(&mut root);
+
+    let root = tree_hash(tx_hashes)?;
+    encode.extend_from_slice(&root.0);
+    // encode.append(&mut root);
+
+    let mut count = serialize(&VarInt(tx_hashes.len() as u64));
     encode.append(&mut count);
-    Ok(hex::encode(encode).into())
+    Ok(hex::encode(encode))
 }
 
 /// Utility function to transform array to fixed array
-pub fn from_slice(bytes: &[u8]) -> [u8; 32] {
+///
+/// ## Panics
+///
+/// Panics if the given slice has a length of less than 32
+pub fn from_slice_to_array(bytes: &[u8]) -> [u8; 32] {
     let mut array = [0; 32];
-    let bytes = &bytes[..array.len()]; // panics if not enough data
+    // TODO: Is a panic wise here? Does every caller know (invariant) the length before calling this function?
```

Comment:
On principle I am totally against panics, as they cause all shorts of issues in incorrect error handling. 
I would personally put an result here and check it. 

If we look at where its called from:
`let expected_merge_mining_hash = Hash::from(from_slice_to_array(&header.merged_mining_hash()));`
This function should not panic has it purely creates a hash, but then again if that fails, I think it returns an empty vec. But the function that calls this, does return a result.

Reply:
I think these conversions can be removed with the other changes I've made anyhow. They are only used in a few places in tests - will remove

Reply:
Removed this conversion :)